[nasa/nos3#580] [nasa/nos3#625] eps system tests and configuration

kevincbruce opened the pull request:
Added System Tests for the EPS based around the System Test Template. Should check all standard things, and confirm basic data points such as charging or discharging based on switch status and In_Sun value. Fixed the EPS charge/discharge model to correctly operate and reflect if it is in or out of the sun. Adjusted the solar panel locations and charge values to closer reflect the size of the STF reference mission, and the bus current values to allow for charging during ambient power loads while in sun.

How to Test:

- Pull the 580 branch, `git submodule update --init --recursive`
- `make clean`, `make`, `make launch`
- Open COSMOS launcher, click the COSMOS button, once windows spawn click Test Runner, then okay on the dialog that pops up
- Go to the `Generic_eps` test suite, and hit run next to the test suite. Wait for it to run successfully.
- Verify in the sim's terminal, while tests are running, that data is changing sensibly based on commands in the tests and that you may manually enter outside of the test (e.g. discharge rate goes up if Switch 0 and/or Switch 1 are enabled and back to ambient if not; verify the solar panel charge rate goes up and down based on the sun vectors and whether you are in sun; etc)
- Try sending the various Sim Bus commands and see if results are reflected (set the battery state of charge low and high and observe the watt hours and voltage change, set a negative rate of charge modifier and observe a discharge, or a positive rate of charge modifier and observe a charge, disable a solar panel that is partially/fully in sun and observe the solar panel charge state go down; enable it again to observe it go back up.
- Try enabling CONUS and Science mode in the manager, as directed in the Science Mode PR, and open telemetry grapher to the `EPS_test.txt` config, and start graphing. See that, once the science mode RTS triggers the instrument, the states of switch 0 and sample both switch on, and that the Voltage starts decreasing.
- During a science pass, try using the Sim Bus Command to set down to ~62 percent. Wait for the voltage to drop below 60% and verify that science mode exits when hitting 60% charge, and that everything is disabled. Then set state of charge to 89% with the same sim bus command, and make sure that when it passes 90% charge science mode reactivates.

Submod Merges:
EPS: https://github.com/nasa-itc/generic_eps/pull/14
Sim_Common: https://github.com/nasa-itc/sim_common/pull/9
COSMOS: https://github.com/nasa-itc/gsw-cosmos/pull/30

Closes #580 #625

dacarter22 committed:
[nasa/nos#580] creating top level branch

kevincbruce committed:
Merge branch 'dev' into nos3#580-eps-system-tests

kevincbruce committed:
[nasa/nos3#580] Added basic test framework for EPS and assured they all work. Need to add proper and in depth device data testing

kevincbruce committed:
[nasa/nos3#580] Updates to EPS data provider, hardware model, and Sim Common to enable charging and discharging based on STF1 panel configuration, an estimate of power per panel for STF1 based on GTOSat values, and whether we are in the sun. Currently, In Sun seems broken despite the fact it should be using the same logic as in Truth 42 Sim, which operates correctly.

kevincbruce committed:
[nasa/nos3#588] fixes to INP_IPC and data provider to add orbit, parse POSN, and calcuate the correct position values. In Sun/Eclipse should be fixed for Solar Panels

kevincbruce committed:
[nasa/nos3#580] Attempted to add Sim Bus commands. Seem nonfunctional currently.

kevincbruce committed:
[nasa/nos3#580] Fixed Sim bus commands; all but state confirmed working; state recognized, but has no effect.

kevincbruce committed:
[nasa/nos3#580] Fixed State of Charge sim command, verified that sim toggle on switch toggle for Sample and Star Tracker also works as intended, though their apps may need changes

kevincbruce committed:
[nasa/nos3#580] Updated system tests to have switches turned off in safeing, and to have the data verification necessary. Tested it during day and eclipse

kevincbruce committed:
Merge branch 'dev' into nos3#580-eps-system-tests

kevincbruce committed:
[nasa/nos3#625] Added EPS Switch toggle commands to Science Mode RTSes; For some reason, they are not loaded/triggering properly?

Isgaroth-the-Green reviewed (approved):
Everything works for me.

kevincbruce committed:
[nasa/nos3#625] Fix to RTS command numbering in SC member size

kevincbruce committed:
[nasa/nos3#625] Added EPS_test.txt config for telemetry grapher for EPS testing. Updated RTSes to fix MIDs for EPS, and to switch around the switch and app disables on RTSes 33, 34, and 35

dacarter22 reviewed (approved):
built and ran with no errors, approving

kevincbruce committed:
[nasa/nos3#625] moved eps test file to be with other tests

kevincbruce committed:
[nasa/nos3#625] updated EPS_test plotter config to be more readable

kevincbruce committed:
[nasa/nos3#625] updated EPS_test plotter config to be more readable; added EPS switch diable to RTS29 (exit science mode)

kevincbruce committed:
[nasa/nos3#625] Update STF_QuickLook.md

Added EPS switch commands to quick look

kevincbruce committed:
[nasa/nos3#625] Added full solar array inhibit which will disconnect solar array from battery at 100% charge and reconnect it at 99% charge, which prevents overcharging

kevincbruce committed:
Merge branch 'nos3#580-eps-system-tests' of github.com:nasa/nos3 into nos3#580-eps-system-tests

kevincbruce commented:
Aside from some watch point changes for Science Mode (which I don't know will be done in this commit or not), everything should be finalized at this point. Now battery will not overcharge when sample is off and it is charging, and everything should be fully integrated with science mode except for those state of charge watchpoints.

kevincbruce committed:
Update STF_QuickLook.md

Updated RTS27 with switch disable, and the LC Watchpoints with the actual MV values for 60% and 90%

kevincbruce committed:
[nasa/nos3#625] updated value to uninhibit solar panels to 99.999% charge instead of 99%; updated EPS_test telemetry graph preset so voltage min and max bonds on plot are set; updated RTS27 to disable Sample Switch on EPS when low battery; updated and enabled watch points for 60% charge science mode exit and 90% charge science mode reentry, and verified those cases work

kevincbruce committed:
Merge branch 'nos3#580-eps-system-tests' of github.com:nasa/nos3 into nos3#580-eps-system-tests

dccutrig committed:
[nasa/nos3#580] Minor update to RTS 26 which wasn't enabling AP 32 previously

Commit:
[nasa/nos3#580] Updates to submodules after PR review;
